CI failing watch tests sometimes

Flaky...

```
=== Failed
=== FAIL: buildengine TestWatch (10.38s)
debug: added module "one"
debug: removed module "one"
debug: added module "one"
    /home/runner/work/ftl/ftl/buildengine/watch_test.go:56: expected at least 4 events, got 3

=== FAIL: buildengine TestWatchWithBuildModifyingFiles (unknown)
debug: removed module "one"
panic: send on closed channel

goroutine 13911 [running]:
github.com/alecthomas/types/pubsub.(*Topic[...]).Publish(...)
	/home/runner/go/pkg/mod/github.com/alecthomas/types@v0.14.0/pubsub/pubsub.go:78
github.com/TBD[54](https://github.com/TBD54566975/ftl/actions/runs/8977809220/job/24657203701?pr=1418#step:6:55)566975/ftl/buildengine.(*Watcher).Watch.func1()
	/home/runner/work/ftl/ftl/buildengine/watch.go:108 +0x7b4
created by github.com/TBD54[56](https://github.com/TBD54566975/ftl/actions/runs/8977809220/job/24657203701?pr=1418#step:6:57)6975/ftl/buildengine.(*Watcher).Watch in goroutine 13909
	/home/runner/work/ftl/ftl/buildengine/watch.go:78 +0x2b1

DONE 308 tests, 1 skipped, 2 failures in 95.391s
Error: === RUN   TestWatch
debug: added module "one"
debug: removed module "one"
debug: added module "one"
    buildengine/watch_test.go:56: expected at least 4 events, got 3
--- FAIL: TestWatch (10.38s)

::error file=	/home/runner/work/ftl/ftl/buildengine/watch.go
	/home/runner/work/ftl/ftl/buildengine/watch.go,line=108 +0x7b4
78 +0x2b1::=== RUN   TestWatchWithBuildModifyingFiles%0Adebug: removed module "one"%0Apanic: send on closed channel%0A%0Agoroutine 13911 [running]:%0Agithub.com/alecthomas/types/pubsub.(*Topic[...]).Publish(...)%0A	/home/runner/go/pkg/mod/github.com/alecthomas/types@v0.14.0/pubsub/pubsub.go:78%0Agithub.com/TBD545[66](https://github.com/TBD54566975/ftl/actions/runs/8977809220/job/24657203701?pr=1418#step:6:67)975/ftl/buildengine.(*Watcher).Watch.func1()%0A	buildengine/watch.go:108 +0x7b4%0Acreated by github.com/TBD5456[69](https://github.com/TBD54566975/ftl/actions/runs/8977809220/job/24657203701?pr=1418#step:6:70)75/ftl/buildengine.(*Watcher).Watch in goroutine 13909%0A	buildengine/watch.go:78 +0x2b1%0A
Error: Process completed with exit code 1.
```

--- a/buildengine/watch_test.go
+++ b/buildengine/watch_test.go
@@ -12,11 +12,11 @@ import (
 	"github.com/alecthomas/types/pubsub"
 
 	. "github.com/TBD54566975/ftl/buildengine"
+	"github.com/TBD54566975/ftl/common/moduleconfig"
 	"github.com/TBD54566975/ftl/internal/log"
 )
 
 const pollFrequency = time.Millisecond * 500
-const halfPollFrequency = time.Millisecond * 250
 
 func TestWatch(t *testing.T) {
 	if testing.Short() {
@@ -29,14 +29,21 @@ func TestWatch(t *testing.T) {
 	w := NewWatcher()
 	events, topic := startWatching(ctx, t, w, dir)
 
-	time.Sleep(pollFrequency + halfPollFrequency) // midway between file scans
+	waitForEvents(t, events, []WatchEvent{})
 
 	// Initiate a bunch of changes.
 	err := ftl("init", "go", dir, "one")
 	assert.NoError(t, err)
 	err = ftl("init", "go", dir, "two")
 	assert.NoError(t, err)
-	time.Sleep(pollFrequency)
+
+	one := loadModule(t, dir, "one")
+	two := loadModule(t, dir, "two")
+
+	waitForEvents(t, events, []WatchEvent{
+		WatchEventProjectAdded{Project: one},
+		WatchEventProjectAdded{Project: two},
+	})
 
 	// Delete a module
 	err = os.RemoveAll(filepath.Join(dir, "two"))
@@ -45,37 +52,11 @@ func TestWatch(t *testing.T) {
 	// Change a module.
 	updateModFile(t, filepath.Join(dir, "one"))
 
-	time.Sleep(pollFrequency)
+	waitForEvents(t, events, []WatchEvent{
+		WatchEventProjectChanged{Project: one},
+		WatchEventProjectRemoved{Project: two},
+	})
 	topic.Close()
-
-	allEvents := []WatchEvent{}
-	for event := range events {
-		allEvents = append(allEvents, event)
-	}
-
-	assert.True(t, len(allEvents) >= 4, "expected at least 4 events, got %d", len(allEvents))
-
-	// Check we've got at least the events we expect.
-	found := 0
-	for _, event := range allEvents {
-		switch event := event.(type) {
-		case WatchEventProjectAdded:
-			if event.Project.Config().Key == "one" || event.Project.Config().Key == "two" {
-				found++
-			}
-
-		case WatchEventProjectRemoved:
-			if event.Project.Config().Key == "two" {
-				found++
-			}
-
-		case WatchEventProjectChanged:
-			if event.Project.Config().Key == "one" {
-				found++
-			}
-		}
-	}
-	assert.True(t, found >= 4, "expected at least 4 events, got %d:\n%v", found, allEvents)
 }
 
 func TestWatchWithBuildModifyingFiles(t *testing.T) {
@@ -94,7 +75,9 @@ func TestWatchWithBuildModifyingFiles(t *testing.T) {
 
 	events, topic := startWatching(ctx, t, w, dir)
 
-	time.Sleep(pollFrequency + halfPollFrequency) // midway between file scans
+	waitForEvents(t, events, []WatchEvent{
+		WatchEventProjectAdded{Project: loadModule(t, dir, "one")},
+	})
 
 	// Change a file in a module, within a transaction
 	transaction := w.GetTransaction(filepath.Join(dir, "one"))
@@ -107,17 +90,8 @@ func TestWatchWithBuildModifyingFiles(t *testing.T) {
 	err = transaction.End()
 	assert.NoError(t, err)
 
-	time.Sleep(pollFrequency)
+	waitForEvents(t, events, []WatchEvent{})
 	topic.Close()
-
-	allEvents := []WatchEvent{}
-	for event := range events {
-		allEvents = append(allEvents, event)
-	}
-	for _, event := range allEvents {
-		event, wasAdded := event.(WatchEventProjectAdded)
-		assert.True(t, wasAdded, "expected only project added events, got %v", event)
-	}
 }
 
 func TestWatchWithBuildAndUserModifyingFiles(t *testing.T) {
@@ -132,10 +106,14 @@ func TestWatchWithBuildAndUserModifyingFiles(t *testing.T) {
 	err := ftl("init", "go", dir, "one")
 	assert.NoError(t, err)
 
+	one := loadModule(t, dir, "one")
+
 	w := NewWatcher()
 	events, topic := startWatching(ctx, t, w, dir)
 
-	time.Sleep(pollFrequency + halfPollFrequency) // midway between file scans
+	waitForEvents(t, events, []WatchEvent{
+		WatchEventProjectAdded{Project: one},
+	})
 
 	// Change a file in a module, within a transaction
 	transaction := w.GetTransaction(filepath.Join(dir, "one"))
@@ -155,27 +133,19 @@ func TestWatchWithBuildAndUserModifyingFiles(t *testing.T) {
 	err = transaction.End()
 	assert.NoError(t, err)
 
-	time.Sleep(pollFrequency)
+	waitForEvents(t, events, []WatchEvent{
+		WatchEventProjectChanged{Project: one},
+	})
 	topic.Close()
+}
 
-	allEvents := []WatchEvent{}
-	for event := range events {
-		allEvents = append(allEvents, event)
+func loadModule(t *testing.T, dir, name string) Module {
+	t.Helper()
+	config, err := moduleconfig.LoadModuleConfig(filepath.Join(dir, name))
+	assert.NoError(t, err)
+	return Module{
+		ModuleConfig: config,
 	}
-	foundChange := false
-	for _, event := range allEvents {
-		switch event := event.(type) {
-		case WatchEventProjectAdded:
-			// expected
-		case WatchEventProjectRemoved:
-			assert.False(t, true, "unexpected project removed event")
-		case WatchEventProjectChanged:
-			if event.Project.Config().Key == "one" {
-				foundChange = true
-			}
-		}
-	}
-	assert.True(t, foundChange, "expected project changed event")
 }
 
 func startWatching(ctx context.Context, t *testing.T, w *Watcher, dir string) (chan WatchEvent, *pubsub.Topic[WatchEvent]) {
@@ -186,6 +156,52 @@ func startWatching(ctx context.Context, t *testing.T, w *Watcher, dir string) (c
 	topic.Subscribe(events)
 
 	return events, topic
+}
+
+// waitForEvents waits for the expected events to be received on the events channel.
+//
+// It always waits for longer than just the expected events to confirm that no other events are received.
+// The expected events are matched by keyForEvent.
+func waitForEvents(t *testing.T, events chan WatchEvent, expected []WatchEvent) {
+	t.Helper()
+	visited := map[string]bool{}
+	expectedKeys := []string{}
+	for _, event := range expected {
+		key := keyForEvent(event)
+		visited[key] = false
+		expectedKeys = append(expectedKeys, key)
+	}
+	eventCount := 0
+	for {
+		select {
+		case actual := <-events:
+			key := keyForEvent(actual)
+			hasVisited, isExpected := visited[key]
+			assert.True(t, isExpected, "unexpected event %v instead of %v", key, expectedKeys)
+			assert.False(t, hasVisited, "duplicate event %v", key)
+			visited[key] = true
+
+			eventCount++
+		case <-time.After(pollFrequency * 5):
+			if eventCount == len(expected) {
+				return
+			}
+			t.Fatalf("timed out waiting for events: %v", visited)
+		}
+	}
+}
+
+func keyForEvent(event WatchEvent) string {
+	switch event := event.(type) {
+	case WatchEventProjectAdded:
+		return "added:" + string(event.Project.Config().Key)
+	case WatchEventProjectRemoved:
+		return "removed:" + string(event.Project.Config().Key)
+	case WatchEventProjectChanged:
+		return "updated:" + string(event.Project.Config().Key)
+	default:
+		panic("unknown event type")
+	}
 }
 
 func ftl(args ...string) error {

